### PR TITLE
Upgraded project config + PEP561 & Mypy + Py3.6+ compatibility + improve CI

### PR DESCRIPTION
**Pull request by demberto**

### CODE REVIEW

**Comment by Thinkeye** [comment]
Hello demberto, thank you for the PR, but I regret to tell you that I have some objections:
- the dataclass itself has been introduced in the Python 3.7, AFAIK, so it doesn't make much sense to make this project compatible with Python 3.6.
- The PEP561 you want to introduce is also available only from Python 3.7 onwards.
- I would also like to keep the first parameter of the decorator as position-only to keep the usage more type safe. There was a reason for PEP 570 in Python 3.8, that's why I have chosen 3.8 to be minimum version for dataclass_struct.

Do you have an use case for the changes you have suggested?

**Comment by demberto** [comment]
I say, forget 3.6 support but retain 3.7 atleast, since 3.7 isn't EOL'd yet unlike 3.6 and since this library is primarily an extension to the dataclasses mechanism wouldn't it make more sense to support it for all the versions dataclasses is in the stdlib?

**Comment by Thinkeye** [comment]
It's interesting topic. In my opinion, the whole point of newer versions of the language is to improve it. New projects usually use newer language versions and older ones are either left unchanged, or upgraded completely. IMHO, there are few, if any projects profiting from extending the compatibility. This project might serve as a small incentive to use Python newer than 3.7. Are you developing/maintaining a project with Python 3.7, which would profit from this change, or do you know about such?

**Comment by demberto** [comment]
Well as a matter of fact, yes I am experimenting this with [PyFLP](https://github.com/demberto/PyFLP) and a new library as well. Well the thing about me is I try to keep compatibility as less as an issue in my libraries. I recommend you to do the same, especially for a small library like this. Especially limiting it to 3.8 just for a very small syntactic sugar?

Its all on you but, at the end people don't like upgrading just because a library doesn't support it, and the likelihood of this library getting used in projects supporting an older version of Python will reduce.

**Comment by Thinkeye** [comment]
Thank you for your insights. If you need the dataclass_struct in 3.7, we can do it. I am not emotional about this, I just want to consider and understand the tradeoff. Would you be so kind to fix the PR, so it passes the GitHub workflows, please?

I would also like to know your rationale for changing the packaging and your reasons for mixing those unrelated changes in a single PR. As the matter of the fact, I would prefer two smaller PRs instead of a single one like this.

File: .github/workflows/python-package.yml [hunk 1]
```diff
@@ -12,11 +12,12 @@ on:
...
   build:
 
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
     strategy:
       fail-fast: false
       matrix:
-        python-version: ["3.8", "3.9", "3.10"]
+        os: ["ubuntu-latest", "windows-latest", "macos-latest"]
+        python-version: ["3.6", "3.7", "3.8", "3.9", "3.10"]
```
Comment (Thinkeye) [inline .github/workflows/python-package.yml:20]: remove version 3.6 as it doesn't have dataclasses yet

Comment (demberto) [inline .github/workflows/python-package.yml:20]: I will do this once my computer gets fixed asap

**Pull request closed without merging**